yapapi/executor/__init__.py

```diff
@@ -205,9 +225,675 @@ def driver(self) -> str:
+                await batch.post()
+                self.emit(events.ScriptFinished(agr_id=agreement_id, task_id=task_id))
+                await self.accept_payment_for_agreement(agreement_id, partial=True)
+                return results
+
+            loop = asyncio.get_event_loop()
+
+            if exec_options.wait_for_results:
+                # Block until the results are available
+                try:
+                    future_results = loop.create_future()
+                    results = await get_batch_results()
+                    future_results.set_result(results)
+                    item = await command_generator.asend(future_results)
+                except StopAsyncIteration:
+                    raise
+                except Exception:
+                    # Raise the exception in `command_generator` (the `worker` coroutine).
+                    # If the client code is able to handle it then we'll proceed with
+                    # subsequent batches. Otherwise the worker finishes with error.
+                    item = await command_generator.athrow(*sys.exc_info())
+            else:
+                # Schedule the coroutine in a separate asyncio task
+                future_results = loop.create_task(get_batch_results())
+                item = await command_generator.asend(future_results)
+
+    async def execute_tasks(
+        self,
+        worker: Callable[
+            [WorkContext, AsyncIterator[Task[D, R]]],
+            AsyncGenerator[Work, Awaitable[List[events.CommandEvent]]],
+        ],
+        data: Union[AsyncIterator[Task[D, R]], Iterable[Task[D, R]]],
+        payload: Payload,
+        max_workers: Optional[int] = None,
+        timeout: Optional[timedelta] = None,
+        budget: Optional[Union[float, Decimal]] = None,
+    ) -> AsyncIterator[Task[D, R]]:
+
+        kwargs: Dict[str, Any] = {"payload": payload}
+        if max_workers:
+            kwargs["max_workers"] = max_workers
+        if timeout:
+            kwargs["timeout"] = timeout
+        kwargs["budget"] = budget if budget is not None else self._budget_amount
+
+        async with Executor(_engine=self, **kwargs) as executor:
+            async for t in executor.submit(worker, data):
+                yield t
+
+
+class Job:
+    """Functionality related to a single job."""
+
+    def __init__(
+        self,
+        engine: Golem,
+        expiration_time: datetime,
+        payload: Payload,
+    ):
+        self.id = str(uuid.uuid4())
+        self.engine = engine
+        self.offers_collected: int = 0
+        self.proposals_confirmed: int = 0
+        self.expiration_time: datetime = expiration_time
+        self.payload: Payload = payload
+
+        self.agreements_pool = AgreementsPool(self.engine.emit)
+        self.finished = asyncio.Event()
+
+    async def _handle_proposal(
+        self,
+        proposal: OfferProposal,
+        demand_builder: DemandBuilder,
```

Comment:
This should probably be an immutable representation of the demand rather than a builder object.

Reply:
seems like a good idea. @azawlocki what do you think?

Reply:
https://github.com/golemfactory/yapapi/issues/380